util/types/time.go

```diff
@@ -333,16 +473,17 @@ func (t *Time) FromPackedUint(packed uint64) error {
 	minute := int((hms >> 6) & (1<<6 - 1))
 	hour := int(hms >> 12)
 
-	nanosec := int(packed%(1<<24)) * 1000
-	err := checkTime(year, month, day, hour, minute, second, nanosec)
+	microsec := int(packed % (1 << 24))
+	err := checkTime(year, month, day, hour, minute, second, microsec*1000)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	loc := local
 	if t.Type == mysql.TypeTimestamp {
 		loc = time.UTC
 	}
-	t.Time = time.Date(year, time.Month(month), day, hour, minute, second, nanosec, loc).In(local)
+	t.Time = FromGoTime(time.Date(year, time.Month(month), day, hour, minute, second, microsec*1000, loc).In(local))
```

Comment:
Because we need convert UTC date to local date.
The whole argument is, notice the `In()`
`time.Date(year, time.Month(month), day, hour, minute, second, microsec*1000, loc).In(local)`